### PR TITLE
Add `mrsolver` tactic, remove old MRSolver interface

### PR DESCRIPTION
**Pull request by m-yac**
This PR removes the `mr_solver_prove`, `mr_solver_test`, etc. commands in lieu of an interface more in line with the rest of SAW. Now there is a tactic `mrsolver` which can be applied to any goal of the form `refinesS ... f1 f2`. (For ease of building refinement goals like this, this PR adds the `refines` command.) For example, here's how the first proof in `sha512_mr_solver.saw` changes with this PR:
```diff
- mr_solver_prove round_00_15 {{ round_00_15_spec }};
+ thm_round_00_15 <-
+   prove_extcore mrsolver (refines [] round_00_15 {{ round_00_15_spec }});
```

Additionally, instead of every goal proved with MRSolver being implicitly added to a global set of rewrites, the user provides to the `mrsolver` tactic a set of the rewrites they wish to apply. This is done using a `Refnset`, inspired heavily by the design of `Simpset`s. For example, here's how the next proof in `sha512_mr_solver.saw` changes with this PR:
```diff
- mr_solver_prove round_16_80 {{ round_16_80_spec }};
+ thm_round_16_80 <-
+   prove_extcore
+     (mrsolver_with (addrefns [thm_round_00_15] empty_rs))
+     (refines [] round_16_80 {{ round_16_80_spec }});
```

These changes are in support of a future development where if MRSolver fails to prove a subgoal, the user will be able to continue the proof script and attempt to prove it themselves. For example, this might look like:
```
thm_round_16_80 <-
  prove_extcore do {
      mrsolver_with (addrefns [thm_round_00_15] empty_rs);
      unfolding [...];
      simplify (...);
      trivial;
    } (refines [] round_16_80 {{ round_16_80_spec }});
```

**Note:** MRSolver currently has some bugs in how it handles Heapster-generated terms caused by changes made last year. Thus most of the Heapster-related MRSolver tests are commented-out – though `mr_solver_unit_tests.saw` still passes.

### CODE REVIEW

File: src/SAWScript/Builtins.hs [hunk 1]
```diff
@@ -2205,15 +2204,25 @@ ensureMonadicTerm sc t
...
+-- how long the function took to run is passed to @strf@ and the result is
+-- printed, then regardless, the last argument is called on the result.
+mrSolverH :: SharedContext ->
+             Maybe SawDoc -> (String -> String) -> Maybe (String -> String) ->
+             (SharedContext -> Prover.MREnv -> Maybe Integer ->
+              (Set VarIndex -> Sequent -> TopLevel (SolverStats, SolveResult)) ->
+              SV.SAWRefnset -> [(LocalName, Term)] -> Term -> Term ->
+              TopLevel (Either Prover.MRFailure a)) ->
+             SV.SAWRefnset -> [(LocalName, Term)] -> TypedTerm -> TypedTerm ->
+             (a -> TopLevel b) -> TopLevel b
```
Comment (bboston7) [inline src/SAWScript/Builtins.hs:2222]: This type signature is very complicated and hard to follow. Does it make sense to convert some of these to records with named fields? If so, that might be easier to understand. If not, then attaching comments to the various arguments in the type signature would help (see [this function](https://github.com/GaloisInc/cryptol/blob/cd5d006277790910f5314092a79a6f15a25ab596/src/Cryptol/Testing/Random.hs#L81) for an example of what I mean)

File: src/SAWScript/Interpreter.hs [hunk 2]
```diff
@@ -3908,52 +3908,49 @@ primitives = Map.fromList
...
+    (funVal2 addrefn)
+    Current
+    [ "Add a proved refinement theorem to a given refinement set." ]
 
-  , prim "mr_solver_set_debug_level" "Int -> TopLevel ()"
-    (pureVal mrSolverSetDebug)
+  , prim "addrefns"            "[Theorem] -> Refnset -> Refnset"
+    (funVal2 addrefns)
+    Current
+    [ "Add proved refinement theorems to a given refinement set." ]
```
Comment (bboston7) [inline src/SAWScript/Interpreter.hs:3937]: Is there any use for refinement sets outside of the experimental Mr Solver commands? If not, then I think `empty_rs`, `addrefn`, and `addrefns` belong in `Experimental`.

File: src/SAWScript/Prover/MRSolver/Evidence.hs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+  -- | An optional annotation, which in the case of SAWScript, is always a
+  -- 'TheoremNonce' indicating from which 'Theorem' this assumption comes
+  fassumpAnnotation :: Maybe t
+}
+
+-- | Recognizes a term of the form:
+-- @(a1:A1) -> ... -> (an:An) -> refinesS_eq ev stack rtp t1 t2@,
+-- and returns a tuple:
+-- @([(a1,A1), ..., (an,An)], ev, ev, stack, stack, rtp, rtp, t1, t2)@
+asRefinesS :: Recognizer Term ([(LocalName, Term)], Term, Term, Term, Term, Term, Term, Term, Term)
```
Comment (bboston7) [inline src/SAWScript/Prover/MRSolver/Evidence.hs:74]: This tuple has enough field that it might be worth refactoring into a record with named fields. Especially because many of them are `Term`s, so it seems easy to accidentally grab the wrong `Term`.

File: src/SAWScript/Prover/MRSolver/Monad.hs [hunk 4]
```diff
@@ -334,104 +353,142 @@ data MRExn = MRExnFailure MRFailure
...
+       Right (a, st) -> return $ Right (a, (mrsSolverStats st, mrsEvidence st))
        Left (MRExnFailure failure) -> return $ Left failure
        Left exn -> fail ("runMRM: unexpected internal exception: " ++ show exn)
 
+-- | Run an 'MRM' computation and return a result or an error, discarding the
+-- final state
+evalMRM :: SharedContext -> Maybe Integer -> MREnv ->
+           (Set VarIndex -> Sequent -> TopLevel (SolverStats, SolveResult)) ->
+           Refnset t -> MRM t a ->
+           TopLevel (Either MRFailure a)
```
Comment (bboston7) [inline src/SAWScript/Prover/MRSolver/Monad.hs:452]: Please document these arguments

File: src/SAWScript/Prover/MRSolver/Monad.hs [hunk 5]
```diff
@@ -334,104 +353,142 @@ data MRExn = MRExnFailure MRFailure
...
+evalMRM sc timeout env askSMT rs =
+  fmap (fmap fst) . runMRM sc timeout env askSMT rs
+
+-- | Run an 'MRM' computation and return a final state or an error, discarding
+-- the result
+execMRM :: SharedContext -> Maybe Integer -> MREnv ->
+           (Set VarIndex -> Sequent -> TopLevel (SolverStats, SolveResult)) ->
+           Refnset t -> MRM t a ->
+           TopLevel (Either MRFailure (SolverStats, MREvidence t))
```
Comment (bboston7) [inline src/SAWScript/Prover/MRSolver/Monad.hs:461]: Add docs for these arguments

File: src/SAWScript/Prover/MRSolver/Solver.hs [hunk 6]
```diff
@@ -1235,67 +1249,65 @@ mrRefinesFunH k _ _ t1 _ t2 = k t1 t2
...
-assumeMRSolver ::
+-- | The continuation passed to 'mrRefinesFunH' in 'refinementTerm' - returns
+-- the 'Term' which is the refinement (@Prelude.refinesS@) of the given
+-- 'Term's, after quantifying over all current 'mrUVars' with Pi types
+refinementTermH :: Term -> Term -> MRM t Term
+refinementTermH t1 t2 =
+  do (SpecMParams ev1 stack1, tp1) <- fromJust . asSpecM <$> mrTypeOf t1
+     (SpecMParams ev2 stack2, tp2) <- fromJust . asSpecM <$> mrTypeOf t2
+     rpre <- liftSC2 scGlobalApply "Prelude.eqPreRel" [ev2, stack2]
+     rpost <- liftSC2 scGlobalApply "Prelude.eqPostRel" [ev2, stack2]
```
Comment (bboston7) [inline src/SAWScript/Prover/MRSolver/Solver.hs:1287]: Just double checking, are these supposed to have the same arguments (`[ev2, stack2]`)? Or is one supposed to have `[ev1, stack1]` as an argument? I'm not familiar enough with the codebase to know, but just wanted to double check that this isn't a copy/paste error.

**Review by bboston7 — APPROVED** [review]
Looks good!  Just a few readability comments to address before merging.

File: src/SAWScript/Prover/MRSolver/Evidence.hs [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+{- |
+Module      : SAWScript.Prover.MRSolver.Evidence
+Copyright   : Galois, Inc. 2023
+License     : BSD3
+Maintainer  : westbrook@galois.com
+Stability   : experimental
+Portability : non-portable (language extensions)
+
+This module defines multiple outward facing components of MRSolver, most
```
Comment (eddywestbrook) [inline src/SAWScript/Prover/MRSolver/Evidence.hs:12]: I feel like the evidence is the least important part of what is currently in this file. Maybe it should be called something else? Also, why is it now separate from Term.hs?

File: src/SAWScript/Builtins.hs (hunk 1 above)
Comment (m-yac) [reply src/SAWScript/Builtins.hs:2222]: Yeah this `mrSolverH` is a mess! I believe my thought process was exactly what you said in your second comment, Eddy: I noticed that I was doing the exact same logic before and after the calls to `Prover.askMRSolver` and `Prover.refinementTerm`, and decided to just abstract all of that out as a separate function. But you make a good point, this would all be easier to understand if this was abstracted out int separate functions which do clearly defined things. I will do that.

File: src/SAWScript/Prover/MRSolver/Solver.hs (hunk 6 above)
Comment (m-yac) [reply src/SAWScript/Prover/MRSolver/Solver.hs:1290]: Good catch! I left some notes on this function that explain: in fact one of the line below the ones you highlighted will assert that `ev1 == ev2` and `stack1 == stack2`, so it actually doesn't matter. I hopefully made that clearer by adding some comments and using `ev2` and `stack2` consistently throughout.

File: src/SAWScript/Prover/MRSolver/Evidence.hs (hunk 7 above)
Comment (m-yac) [reply src/SAWScript/Prover/MRSolver/Evidence.hs:12]: I think my motivation to make this separate from `Term.hs` is generally because I prefer smaller, easier to digest Haskell files over larger files containing many separate linked ideas. This set of definitions seem like a linked idea, so I decide to break them out – though I realized I never actually explained how they were linked. I added a comment at the top now but here it is reproduced here:
```
Note: In order to avoid circular dependencies, the 'FunAssump' type and its
dependents in this file ('Refnset' and 'MREvidence') are given a type
parameter `t` which in practice always be 'TheoremNonce' from `Value.hs`.
The reason we cannot just import `Value.hs` here directly is because the
'Refnset' type is used in `Value.hs` - specifically, in the 'VRefnset'
constructor of the 'Value' datatype.
```

I am not sold on the name `Evidence.hs`, though. My thoughts were: (1) at some point that `Evidence` datatype *will* needed to be fleshed out, at which point that section of the file will get much bigger, and (2) the `FunAssump` and `Refnset` datatypes both share the type parameter described above and are both used as key components in `Evidence`, so `Evidence` might be a good way to center the file. But, (1) is not the case now and for (2), you could make the argument that `FunAssump` or `Refnset` are just as central.

So as for the name, up to you. If you think strongly it should be merged back into `Term.hs` that's fine too, I'm just worried about these files having too many parts which can make them harder to read.

File: src/SAWScript/Builtins.hs (hunk 1 above)
Comment (m-yac) [reply src/SAWScript/Builtins.hs]: Okay I broke up this function into smaller pieces, let me know what you guys think!

Comment (bboston7) [reply src/SAWScript/Builtins.hs]: Much better! Thanks!
